Reduce the size of code generated by layer macros

joroKr21 opened the pull request:
Avoid pruning intermediate results.
Only user-provided layers are pruned.
Switch to implicit trace.

kyri-petrou commented on core/shared/src/main/scala-3/zio/internal/macros/LayerMacroUtils.scala:
```diff
@@ -42,31 +42,27 @@ private[zio] object LayerMacroUtils {
 
     '{
-      val trace: Trace = Tracer.newTrace
+      given Trace = Tracer.newTrace
```
Just for my own understanding, what's the rationale behind this change? AFAIK `given` generate more bytecode than plain `val`s.

joroKr21 replied:
Why do you think so? The change - it's cleaner code and also making sure we don't omit it somewhere.

kyri-petrou replied:
`given`s in Scala 3 without type parameters or implicit arguments are the equivalent of `implicit lazy val` in Scala 2. `lazy val`s contain a fair bit of byte code because they need to guarantee _at most once_ semantics even when concurrently accessed by multiple threads. One way to reduce the byte-code generated by `given`s is to annotate them with `@threadUnsafe` so that they use an implementation that is faster/lighter but not thread-safe.
Another option is to use either `implicit val` or `implicit def`.

As an example, given the following code:

```scala
//> using scala 3.5.1
//> using dep "dev.zio::zio::2.1.11"
//> using options -Xprint:genBCode

import zio.Trace
import zio.internal.stacktracer.Tracer

import scala.annotation.threadUnsafe

object Foo {
  given trace1: Trace               = Tracer.newTrace
  @threadUnsafe given trace2: Trace = Tracer.newTrace
  implicit val trace3: Trace        = Tracer.newTrace
}
```

<details>
  <summary>This is what the Scala 3 compiler generates (Click to expand)</summary>
```scala
[[syntax trees at end of                  genBCode]] // /Users/kpe/Repos/sandbox/test-scala-cli/src/testy.scala
package <empty> {
  @SourceFile("src/testy.scala") final module class Foo extends Object {
    def <init>(): Unit =
      {
        super()
        Foo.trace3 = "<empty>.Foo.trace3(testy.scala:13)":Object
        ()
      }
    @static private def <clinit>(): Unit =
      {
        Foo.OFFSET$_m_0 =
          scala.runtime.LazyVals.getOffsetStatic(
            classOf[Object {...}].getDeclaredField("trace1$lzy1"))
        ()
      }
    @static @static <static> val OFFSET$_m_0: Long =
      scala.runtime.LazyVals.getOffsetStatic(
        classOf[Object {...}].getDeclaredField("trace1$lzy1"))
    private def writeReplace(): Object =
      new scala.runtime.ModuleSerializationProxy(classOf[Foo])
    @volatile private final lazy given var trace1$lzy1: Object = null
    final lazy given def trace1(): Object =
      {
        val result: Object = Foo.trace1$lzy1
        if 
          result.!=(null).&&(
            result.isInstanceOf[
              scala.runtime.LazyVals.LazyVals$LazyValControlState].unary_!()
          )
         then result else 
          if result.eq(scala.runtime.LazyVals.NullValue) then null else 
            Foo.trace1$lzyINIT1()
      }
    private def trace1$lzyINIT1(): Object =
      while <empty> do 
        {
          val current: Object = Foo.trace1$lzy1
          if current.eq(null) then
            if 
              scala.runtime.LazyVals.objCAS(this, Foo.OFFSET$_m_0, null,
                scala.runtime.LazyVals.Evaluating)
             then
              {
                var result: Object = null
                var resultNullable: Object = null
                try 
                  {
                    resultNullable = "<empty>.Foo.trace1(testy.scala:11)":Object
                    if resultNullable.eq(null) then
                      result = scala.runtime.LazyVals.NullValue else 
                      result = resultNullable
                    ()
                  }
                 finally 
                  if 
                    scala.runtime.LazyVals.objCAS(this, Foo.OFFSET$_m_0,
                      scala.runtime.LazyVals.Evaluating, result).unary_!()
                   then
                    {
                      val lock: scala.runtime.LazyVals.LazyVals$Waiting =
                        Foo.trace1$lzy1.asInstanceOf[
                          scala.runtime.LazyVals.LazyVals$Waiting]
                      scala.runtime.LazyVals.objCAS(this, Foo.OFFSET$_m_0,
                        lock, result)
                      lock.countDown()
                    }
                   else ()
                return resultNullable
              }
             else ()
           else 
            if 
              current.isInstanceOf[
                scala.runtime.LazyVals.LazyVals$LazyValControlState]
             then
              if current.eq(scala.runtime.LazyVals.Evaluating) then
                {
                  scala.runtime.LazyVals.objCAS(this, Foo.OFFSET$_m_0, current,
                    new scala.runtime.LazyVals.LazyVals$Waiting())
                  ()
                }
               else 
                if current.isInstanceOf[scala.runtime.LazyVals.LazyVals$Waiting]
                   then
                  current.asInstanceOf[scala.runtime.LazyVals.LazyVals$Waiting].
                    await()
                 else return null
             else return current
        }
    private final lazy given var trace2$lzy1: Object = null.asInstanceOf[Object]
    private lazy var trace2bitmap$1: Boolean = false
    @threadUnsafe final lazy given def trace2(): Object =
      {
        if Foo.trace2bitmap$1.unary_!() then
          {
            Foo.trace2$lzy1 = "<empty>.Foo.trace2(testy.scala:12)":Object
            Foo.trace2bitmap$1 = true
          }
         else ()
        Foo.trace2$lzy1
      }
    private <static> val trace3: Object
    implicit def trace3(): Object = Foo.trace3
  }
  final lazy module val Foo: Foo = new Foo()
}
```
</details>

So in order to keep the macro-generated code as lean as possible, let's use either `implicit val` or the `@threadUnsafe` annotation (which should be OK to use here)

kyri-petrou commented on core/shared/src/main/scala-3/zio/internal/macros/LayerMacroUtils.scala:
```diff
@@ -42,31 +42,27 @@ private[zio] object LayerMacroUtils {
-                      report.errorAndAbort(s"Cannot find EnvironmentTag[${TypeRepr.of[o2].show}] in implicit scope")
-                    )
-                  '{ $lhs.++($rhs)($tag) }
+                  rhs.asTerm match {
+                    case _: Ident => '{ $lhs.and($rhs)(summonInline) }
+                    case _        => '{ $lhs +!+ $rhs }
+                  }
```
I'm not too familiar with `Ident`. How exactly does it differentiate between user defined code from macro-generated one?

Also, can we use `++` instead of `and` for consistency?

joroKr21 replied:
`Ident` is short for "identifier" - that's a reference to a value as opposed to a more complex expression (composed layer in this case). We can use `++`, just a matter of style I guess but I'm not a fan of calling a symbolic method with dot notation.

joroKr21 commented:
CI is flaky

kyri-petrou reviewed (commented):
LGTM, although as a sanity check, could you publish your PR locally and ensure that we're not reintroducing this issue https://github.com/zio/zio/issues/8986?

We should also probably figure out a way to add a test for this during CI but not really sure how :/

joroKr21 committed:
Reduce the size of code generated by layer macros

Avoid pruning intermediate results.
Only user-provided layers are pruned.
Switch to implicit trace.

joroKr21 commented:
> LGTM, although as a sanity check, could you publish your PR locally and ensure that we're not reintroducing this issue https://github.com/zio/zio/issues/8986?

Indeed, I reverted to checking with boolean variables. Unfortunately, there's no easy way to know if an implicit is used or not before type checking.

> We should also probably figure out a way to add a test for this during CI but not really sure how :/

Not sure about CI but you can check it in the `sbt console`, you don't need to publish.
